facebook/hermes
# Rework iOS & Mac CI setup to fix creating artefacts

There are a few things that I have reworked in this PR, and I have done my best to explain them in as much details as possible.

## The challenge

First, let me start with the root cause of why I started working on this PR:

> The `hermes-engine-darwin-vX.X.X.tgz` archive created by CircleCI (and later published to npm or CocoaPods) contains iPhone only architectures, and does not contain MacOS and iPhone Simulator ones.

The reason for that is simple 👇

On CircleCI, we use [`cmake` to build a special `hermes-runtime-darwin-cocoapods-release` target](https://github.com/facebook/hermes/blob/master/.circleci/config.yml#L292). In order to actually build it, we need to tell `cmake` the build folder to use.

The challenge is that we have three separate build folders - `build_iphoneos`, `build_iphonesimulator` and `build_macosx` - each, for a different platform that we support. Thanks to that, we can avoid different issues (especially between iPhone/iPhone Simulator and Mac) when building different platforms back and forth, as well as share the cache on subsequent builds. We can be also more explicit when invoking `cmake`, which helps to locate the correct build folder, e.g. [when building CLI tools](https://github.com/facebook/hermes/pull/393/commits/8c3f2fe793e6fc40b46d81fcafa2dcf6161c4e3b).

Side effect of that design decision is that **we can't just use `cmake` to build an archive with all platforms and architectures we want**. It just doesn't work (or I am not aware of a way to make it work).

## The solution

Do not build a custom `cmake` target (named "hermes-runtime-darwin-cocoapods-release"). Use plain bash to build Hermes for each platform we want and then, manually copy and package that into an archive. 

We already use that same approach while in development - there is a custom post install phase in `hermes-engine.podspec` It consists of `build_apple_framework` that builds Hermes for a given platform with flags/settings specified and `create_universal_framework` that combines multiple platforms into a single framework (e.g. iPhone and iPhone Simulator).

## Additional benefits

Together with aforementioned, I have implemented several additional improvements.

1. There is no `spec.prepare_command` in a Hermes podspec

We no longer implicitly build a framework when running `pod install` in a `test/ApplePlatformsIntegrationTestApp`. We now need to explicitly run e.g. `./utils/build-mac-framework.sh`.

As a result, I have decided to remove that step from a podspec. We no longer need to conditionally include/remove it, depending on a presence of an environmental variable.

2. Better CircleCI configuration with shared cache

Since scripts were extracted to separate files, they can be now directly run from a CircleCI. We don't need a custom `cmake` target, we can use the same code that is already used in development (and has been tested to work flawlessly).

Thanks to a few different scripts, we can have a set of CircleCI jobs, each responsible for a different aspect of building and packaging Hermes for Darwin:
- build_apple_platforms - to build MacOS, iPhone, iPhone Simulator and CLI
- test_apple_platforms - to run tests on previously built frameworks
- package_apple_platforms - to package frameworks into archives

The build action is performed only once, as a part of `build_apple_platforms` and is cached for subsequent runs and following jobs in the pipeline, to reduce the amount of time needed to test and verify results.

## How to test it

Copy `Hermes-engine-darwin-vx.x.x.tar.gz` from CircleCI `npm` job and paste it into your `Pods` folder. It will replace currently released version to CocoaPods that doesn't contain all the platforms.

Re-run on iPhone Simulator. Linking error with missing files/classes for given architecture will be gone.

--

Fixes https://github.com/facebook/hermes/issues/353 (we no longer build CLI separately).
